replaced select component with autocomplete

KK4ito opened the pull request:
for #2419

- add validation to allow only values from list
- add new interface type: platform
- autoselect value if input matches with an entry from list

dtslvr commented on apps/client/src/app/pages/accounts/create-or-update-account-dialog/create-or-update-account-dialog.component.ts:
```diff
@@ -4,14 +4,24 @@ import {
-import { CreateOrUpdateAccountDialogParams } from './interfaces/interfaces';
+import {
+  CreateOrUpdateAccountDialogParams,
+  Platform
```
Can you reuse `import { Platform } from '@prisma/client';` instead of your own interface?

KK4ito replied:
yes sure, bd2b36622d4ccea0b9086ba6c7ba4f66be2010b3

because of the property "url" i had to map the infos from
`this.dataService.fetchInfo();`

with the following adjustment

```
 this.platforms = platforms.map((platform) => {
      return { ...platform, url: '' };
    });
```

dtslvr commented on apps/client/src/app/pages/accounts/create-or-update-account-dialog/create-or-update-account-dialog.component.ts:
```diff
@@ -23,7 +33,8 @@ import { CreateOrUpdateAccountDialogParams } from './interfaces/interfaces';
   public currencies: string[] = [];
-  public platforms: { id: string; name: string }[];
+  public platforms: Platform[];
+  public filteredPlatforms: Observable<Platform[]>;
```
Can you move up ⬆️ to keep the alphabetical order?

KK4ito replied:
done bd2b36622d4ccea0b9086ba6c7ba4f66be2010b3

dtslvr commented on apps/client/src/app/pages/accounts/create-or-update-account-dialog/create-or-update-account-dialog.component.ts:
```diff
@@ -47,8 +58,22 @@ export class CreateOrUpdateAccountDialog implements OnDestroy {
-      platformId: [this.data.account.platformId]
+      platformId: [
+        this.platforms.find(
+          (platform) => platform.id === this.data.account.platformId
```
Can you use the more explicit form for consistency?

```
({id}) => { return id === this.data.account.platformId; }
```

KK4ito replied:
thanks for the feedback, didnt know about the explicit form :D quite a long time since my last work with typescript

dtslvr commented on apps/client/src/app/pages/accounts/create-or-update-account-dialog/create-or-update-account-dialog.component.ts:
```diff
@@ -79,4 +103,33 @@ export class CreateOrUpdateAccountDialog implements OnDestroy {
+  public autoCompleteCheck() {
+    const inputValue = this.accountForm.controls['platformId'].value;
+    const matchingEntry = this.platforms.find(
+      (platform) => platform.name === inputValue
```
`({ name }) => { return name === inputValue; }`

KK4ito replied:
during testing, I encountered a bug with my previous implementation. I fixed it by checking the input value.

only for inputvalue of type string do the stuff (find, setting value)

inputValue can be of type platform, when a value from the list is selected by user interaction

KK4ito replied:
additional thoughts: this approach (matching by platform name) only works if we can guarantee that there are no multiple platforms with the same name.
or else the current implementation returns the first match from the given array

dtslvr commented on apps/client/src/app/pages/accounts/create-or-update-account-dialog/create-or-update-account-dialog.component.ts:
```diff
@@ -79,4 +103,33 @@ export class CreateOrUpdateAccountDialog implements OnDestroy {
+  }
+
+  displayFn(platform: Platform) {
+    return platform && platform.name ? platform.name : '';
```
`return platform?.name ? platform.name : '';`

dtslvr commented on apps/client/src/app/pages/accounts/create-or-update-account-dialog/create-or-update-account-dialog.component.ts:
```diff
@@ -79,4 +103,33 @@ export class CreateOrUpdateAccountDialog implements OnDestroy {
+    const filterValue = value.toLowerCase();
+
+    return this.platforms.filter((platformEntry) =>
+      platformEntry.name.toLowerCase().includes(filterValue)
```
`{ return name.toLowerCase().includes(filterValue); }`

dtslvr commented on apps/client/src/app/pages/accounts/create-or-update-account-dialog/create-or-update-account-dialog.html:
```diff
@@ -33,12 +33,23 @@ <h1 *ngIf="!data.account.id" i18n mat-dialog-title>Add account</h1>
+            [value]="platformEntry"
           >
-        </mat-select>
+            {{platformEntry.name}}
```
`{{ platformEntry.name }}`

dtslvr commented on apps/client/src/app/pages/accounts/create-or-update-account-dialog/create-or-update-account-dialog.module.ts:
```diff
@@ -21,6 +22,7 @@ import { CreateOrUpdateAccountDialog } from './create-or-update-account-dialog.c
     MatFormFieldModule,
     MatInputModule,
     MatSelectModule,
+    MatAutocompleteModule,
```
Can you move up ⬆️  to keep the alphabetical order?

dtslvr reviewed (changes requested):
Cool 😎  I have added some comments.

KK4ito commented:
thanks for the fast review. I applied all suggestions,

KK4ito commented:
> Thank you very much for the changes, @KK4ito!
> 
> I have played around a bit and added some more comments. It happened to me several times that I clicked a search result, but the value was not applied. I can't reproduce it reliably and unfortunately there is no error in the console. Can you take a closer look?

Thank you for the feedback.
I tested it and I think I was be able to reproduce & fix it. This behaviour occured when there is a input in the text field during user click. In such cases the input value was from type string and the value was forcibly set to undefined. 94a99d188819fe371abc7649b75890e2c204e69a

KK4ito commented on apps/client/src/app/pages/accounts/create-or-update-account-dialog/create-or-update-account-dialog.component.ts:
```diff
@@ -79,4 +103,34 @@ export class CreateOrUpdateAccountDialog implements OnDestroy {
+    }
+  }
+
+  displayFn(platform: Platform) {
```
done

KK4ito commented on apps/client/src/app/pages/accounts/create-or-update-account-dialog/create-or-update-account-dialog.component.ts:
```diff
@@ -79,4 +103,34 @@ export class CreateOrUpdateAccountDialog implements OnDestroy {
+  }
+
+  private _autocompleteObjectValidator(): ValidatorFn {
+    return (control: AbstractControl): { [key: string]: any } | null => {
```
done

KK4ito commented on apps/client/src/app/pages/accounts/create-or-update-account-dialog/create-or-update-account-dialog.component.ts:
```diff
@@ -79,4 +103,34 @@ export class CreateOrUpdateAccountDialog implements OnDestroy {
+  private _filter(value: string): Platform[] {
+    const filterValue = value.toLowerCase();
+    return this.platforms.filter(({ name }) => {
+      return name.toLowerCase().includes(filterValue);
```
done

dtslvr commented on apps/client/src/app/pages/accounts/create-or-update-account-dialog/create-or-update-account-dialog.component.ts:
```diff
@@ -79,4 +101,36 @@ export class CreateOrUpdateAccountDialog implements OnDestroy {
+  }
+
+  public displayFn(platform: Platform) {
+    return platform?.name ? platform.name : '';
```
`return platform?.name ?? '';`

dtslvr reviewed (changes requested):
I have simplified the platforms (type) a bit. Please fetch my update.

Unfortunately, I noticed one more problem: When I clear the platform, the form component gets red and the form is invalid. Do you have an idea?

KK4ito commented:
> I have simplified the platforms (type) a bit. Please fetch my update.
> 
> Unfortunately, I noticed one more problem: When I clear the platform, the form component gets red and the form is invalid. Do you have an idea?

Thanks for testing and feedback. This error happens because of the autocompelteObjectValidator. Fixed it with 473d466656e436977cd21241a2856bd03912ee5d

dtslvr reviewed (approved):
Had to fix an issue with removing a platform when you edit an account. Now I think it works fine. Thanks a lot for your contribution, Kevin! 😎
